osf/models/mixins.py

```diff
@@ -459,3 +465,59 @@ def get_extra_log_params(self, comment):
         """Return extra data to pass as `params` to `Node.add_log` when a new comment is
         created, edited, deleted or restored."""
         return {}
+
+class TaxonomizableMixin(models.Model):
+
+    class Meta:
+        abstract = True
+
+    subjects = models.ManyToManyField(blank=True, to='osf.Subject', related_name='%(class)ss')
+
+    @cached_property
+    def subject_hierarchy(self):
+        return [
+            s.object_hierarchy for s in self.subjects.exclude(children__in=self.subjects.all())
+        ]
+
+    def set_subjects(self, preprint_subjects, auth, add_log=True):
+        if self.is_registration:
+            raise PermissionsError('Registrations may not be modified.')
+        if self.is_collection:
+            raise NodeStateError('Collections may not have subjects')
+        if not self.has_permission(auth.user, ADMIN):
+            raise PermissionsError('Only admins can change subjects.')
+
+        old_subjects = list(self.subjects.values_list('id', flat=True))
```

Comment:
Keeping a QS in memory causes the query to be executed each evaluation. When the underlying data changes, so does the result of the evaluation -- meaning it is necessary here.
```
In [8]: olds1 = p.subjects  # .copy isn't a SubjectQuerySet method

In [9]: olds2 = p.subjects.all()

In [10]: p.subjects.count()
Out[10]: 4

In [11]: olds2.count()
Out[11]: 4

In [12]: olds1.count()
Out[12]: 4

In [13]: p.subjects.clear()

In [14]: p.subjects.count()
Out[14]: 0

In [15]: olds1.count()
Out[15]: 0

In [16]: olds2.count()
Out[16]: 0
```
You could work around this with something like 
```
old_sub_qs = Subject.objects.filter(id__in=list(self.subjects.values_list('id', flat=True)))
```
But this doesn't save you from making queries on subsequent evaluations:
```
In [2]: old_sub_qs = Subject.objects.filter(id__in=list(p.subjects.values_list('id', flat=True)))

In [3]: from django.test.utils import CaptureQueriesContext

In [4]: from django.db import connection

In [5]: with CaptureQueriesContext(connection) as ctx0:
   ...:     list(old_sub_qs.all())
   ...:

In [6]: ctx0.final_queries
Out[6]: 4

In [7]: ctx0.final_queries - ctx0.initial_queries
Out[7]: 1

In [8]: with CaptureQueriesContext(connection) as ctx1:
   ...:     list(old_sub_qs.values('id', 'text'))
   ...:

In [9]: ctx1.final_queries
Out[9]: 5

In [10]: ctx1.final_queries - ctx1.initial_queries
Out[10]: 1

```
In order to avoid that, we'd need to evaluate `old_subjects` to a list of objects and pass that around instead, using comprehensions to extract relevant data. Given the (small) extra memory cost, and the fact that it would only save us a couple queries on a small / relatively bounded table, I'm going to say this isn't worth it.